packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -123,9 +123,84 @@ class LineBoundary extends TextBoundary {
   TextRange getTextBoundaryAt(int position) => _textLayout.getLineAtOffset(TextPosition(offset: max(position, 0)));
 }
 
+/// A text boundary that uses paragraphs as logical boundaries.
+///
+/// A paragraph is defined as the range between line terminators. If no
+/// line terminators exist then the paragraph boundary is the entire document.
+class ParagraphBoundary extends TextBoundary {
+  /// Creates a [ParagraphBoundary] with the text.
+  const ParagraphBoundary(this._text);
+
+  final String _text;
+
+  /// Returns the [int] representing the start position of the paragraph that
+  /// bounds the given `position`. The returned [int] is the position of the code unit
+  /// that follows the line terminator that encloses the desired paragraph.
+  @override
+  int? getLeadingTextBoundaryAt(int position) {
+    if (position < 0 || position > _text.length) {
+      return null;
+    }
+    if (position == 0) {
```

Comment:
I'm unsure if we should be returning `null` or `0` in the case (I'm leaning towards 0 since that would have `getTextBoundaryAt` return a collapsed TextRange at 0). Do you have any opinions on this?

Reply:
I went with `0` for the reason above.